### PR TITLE
Improve handling of not supported environments ex. database versions

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue #38694

### Summary of Changes

As mentiond here #38694 there is no transparent messaging from Joomla when you are running an site which does not meet the minimum requirements for example of the database.

I see here two solutions to this:
- improve the messaging within the CMS
- had bail out of the CMS and show a message similiar to the PHP minimum version

The latter is not b/c and if anything only something for majors to be changed.

### Testing Instructions

- Setup a site that is running MySQL 5.5
- Install Joomla 4.2.0
- Note that an update to 4.2.2 or later will be shown
- But no download URL will be found
- Also check the installer warings (System -> Warnings)
- Confirm that there is no warning showin by the core about that database version issue
- apply this patch
- Check the "We cant find an update" message again it will mention an (to be created once agreed here) doc page instead of a message which does not help to solve the issue.
- Check the warnings tab and confirm there is a clear message showing the issue with the minimum database version.

#### When you dont have a mysql 5.5 or older setup

In that case you can fake the tests mentiond here by doing the following:

Add the following two lines to this file: `administrator\components\com_joomlaupdate\src\View\Joomlaupdate\HtmlView.php` at Line 179-180
```php
        $hasUpdate = true;
        $hasDownload = false;
```

This will make sure you get the "no download URL found all the time".

The seccond check (new warning message) can be faked by reversing the check here after applying the patch: `administrator\components\com_installer\src\Model\WarningsModel.php` at Line 180

from
```php
if (Factory::getDbo()->isMinimumVersion() === false) {
```
to
```php
if (Factory::getDbo()->isMinimumVersion() === true) {
```

#### We cant find a download URL message
Screenshot 2022-09-21 140628

#### Warning within the com_installer warnings

No message at all.

#### We cant find a download URL message
image

#### Warning within the com_installer warnings
image

### Documentation Changes Required

The new page dedicated to this issue has to be created maybe under the "Help pages space"?: https://docs.joomla.org/Special:MyLanguage/J4.x:We_cant_find_a_download_url

@coolwebcreations Do you have a preference where to place such new doc page?

Said page should list the possible reasons for this message:
- The update to Joomla %1$s is not available for your stability level (where to find and change that setting)
- the update server setting (index.php?option=com_config&view=component&component=com_joomlaupdate) is not pointing to a official updateserver (where to find and change that setting)
- this environment does not meet the minimum requirements (https://downloads.joomla.org/technical-requirements) for this update (where to find and change that setting)
- there is a problem with the Joomla Update Server. (where to report issues with the update server)

And also mention more deeper ways to check the points mentiond above. Like with the new warning within com_installer as well as that the update can be manually installed. But that should always be the last resort anyway.

By this we keep the message within the CMS as short as possibe but also have a dedicated place to document the issues and solutions.

### CODE REVIEW

File: administrator/language/en-GB/com_installer.ini [hunk 1]
```diff
@@ -204,6 +204,8 @@ COM_INSTALLER_MSG_WARNINGS_UPDATE_NOTICE="Before updating ensure that the update
...
 COM_INSTALLER_MSG_WARNINGS_UPLOADFILETOOBIG="The selected file cannot be uploaded as it is bigger than the maximum upload size."
+COM_INSTALLER_MSG_WARNINGS_NOTSUPPORTEDDATABASEVERSIONDESC="The version of the database system you are using does not meet the <a href=\"https://downloads.joomla.org/technical-requirements\" class=\"alert-link\" target=\"_blank\" rel=\"noopener noreferrer\">core minimum requirements</a>. Right now you are running the following configuration:<br><ul><li>Database type: %1$s </li><li>Database version: %2$s</li><li>Minimum database version: %3$s</li></ul>This means you are running Joomla within a not tested and not-supported environment which can result into issues with the core and extensions. This will also mean that Joomla will not offer any updates via the Joomla Updater to your installation."
```
Comment (brianteeman) [inline administrator/language/en-GB/com_installer.ini]:
```suggestion
COM_INSTALLER_MSG_WARNINGS_NOTSUPPORTEDDATABASEVERSIONDESC="The version of the database system you are using does not meet the <a href=\"https://downloads.joomla.org/technical-requirements\" class=\"alert-link\" target=\"_blank\" rel=\"noopener noreferrer\">core minimum requirements</a>. Your current configuration is:<br><ul><li>Database type: %1$s </li><li>Database version: %2$s</li><li>Minimum database version: %3$s</li></ul>This means you are running Joomla on an untested and unsupported environment which may result in issues with the core and extensions. Joomla will not provide any updates until your database is upgraded to reach at least the minimum supported version."
```

File: administrator/language/en-GB/com_installer.ini [hunk 2]
```diff
@@ -204,6 +204,8 @@ COM_INSTALLER_MSG_WARNINGS_UPDATE_NOTICE="Before updating ensure that the update
...
+COM_INSTALLER_MSG_WARNINGS_NOTSUPPORTEDDATABASEVERSION="The current database version does not meet the minimum requirements"
```
Comment (brianteeman) [inline administrator/language/en-GB/com_installer.ini]:
```suggestion
COM_INSTALLER_MSG_WARNINGS_NOTSUPPORTEDDATABASEVERSION="Database does not meet the minimum version requirements."
```

File: administrator/language/en-GB/com_joomlaupdate.ini [hunk 3]
```diff
@@ -43,7 +43,7 @@ COM_JOOMLAUPDATE_MINIMUM_STABILITY_RC="Release Candidate"
 COM_JOOMLAUPDATE_MINIMUM_STABILITY_STABLE="Stable"
 COM_JOOMLAUPDATE_NODOWNLOAD_EMPTYSTATE_APPEND="Upload and Update"
 COM_JOOMLAUPDATE_NODOWNLOAD_EMPTYSTATE_BUTTON_ADD="Retry check for update"
-COM_JOOMLAUPDATE_NODOWNLOAD_EMPTYSTATE_CONTENT="An update to Joomla %1$s was found, but it wasn't possible to fetch the download URL for that update. Either the update to Joomla %1$s is not available for your stability level or there is a problem with the Joomla Update Server.<br>Please try to download the update package from <a href=\"https://downloads.joomla.org/latest\">the official Joomla download page</a> and use the Upload and Update function."
+COM_JOOMLAUPDATE_NODOWNLOAD_EMPTYSTATE_CONTENT="An update to Joomla %1$s was found, but it wasn't possible to fetch the download URL for that update. This can have different reasons please check this <a href=\"https://docs.joomla.org/Special:MyLanguage/J4.x:We_cant_find_a_download_url\" class=\"alert-link\" target=\"_blank\" rel=\"noopener noreferrer\">help page</a> for more details."
```
Comment (brianteeman) [inline administrator/language/en-GB/com_joomlaupdate.ini]:
```suggestion
COM_JOOMLAUPDATE_NODOWNLOAD_EMPTYSTATE_CONTENT="An update to Joomla %1$s was found, but it wasn't possible to fetch the download URL for the update. There can be different reasons for this. Please check this <a href=\"https://docs.joomla.org/Special:MyLanguage/J4.x:We_cant_find_a_download_url\" class=\"alert-link\" target=\"_blank\" rel=\"noopener noreferrer\">help page</a> for more details."
```

**Review by zero-24 — COMMENTED** [review]
Thanks @brianteeman

File: administrator/components/com_installer/src/Model/WarningsModel.php [hunk 4]
```diff
@@ -177,6 +177,19 @@ public function getItems()
             ];
         }
 
+        if (Factory::getDbo()->isMinimumVersion() === false) {
```
Comment (fancyFranci) [inline administrator/components/com_installer/src/Model/WarningsModel.php]: Can you use `$this->getDatabase()` instead of `Factory::getDbo()` ?

Comment (zero-24) [reply administrator/components/com_installer/src/Model/WarningsModel.php]:
```suggestion
        if ($this->getDatabase()->isMinimumVersion() === false) {
```

Comment (zero-24) [reply administrator/components/com_installer/src/Model/WarningsModel.php]: Done with the last commit thanks

File: administrator/components/com_installer/src/Model/WarningsModel.php [hunk 5]
```diff
@@ -177,6 +177,19 @@ public function getItems()
             ];
         }
 
+        if ($this->getDatabase()->isMinimumVersion() === false) {
+            $db         = Factory::getDbo();
```
Comment (laoneo) [inline administrator/components/com_installer/src/Model/WarningsModel.php]: I think here should also `$this->getDatabase()` be used.

Comment (zero-24) [reply administrator/components/com_installer/src/Model/WarningsModel.php]:
```suggestion
            $db         = $this->getDatabase();
```

Comment (zero-24) [reply administrator/components/com_installer/src/Model/WarningsModel.php]: Yes done

**Comment by fancyFranci** [comment]
It is a useful improvement of the updater. Can you test this please @joeforjoomla ? :)

**Comment by coolcat-creations** [comment]
The Update Message does not seem to be quite right. It says that an update was found which is not possible I guess, because there is no update?
grafik

**Comment by zero-24** [comment]
@coolcat-creations please click "check for Updates" thats not the message that is changed here.

**Comment by coolcat-creations** [comment]
I clicked on Check for Updates and thats the message (like in the screenshot) :-/

**Comment by webgras** [comment]
I have a client site with MySQL 5.1, I have the problem.
(However, I could not user patchtester because it is not installable on this hosting for mysql 5.1 (I guess), so I changed the files manually.)
Now my screens look like this:
grafik
grafik

For me it is correct, because there is an official Update for 4.2.7
and the warning is with valid information.
So, I think this is a confirmed test and expected behaviour, right?

**Comment by zero-24** [comment]
> So, I think this is a confirmed test and expected behaviour, right?

Yes please record it via the issue Tracker.

I have just updated the branch to make it ready to be merged by the release Leads :)

File: administrator/language/en-GB/com_installer.ini [hunk 6]
```diff
@@ -204,6 +204,8 @@ COM_INSTALLER_MSG_WARNINGS_UPDATE_NOTICE="Before updating ensure that the update
...
 COM_INSTALLER_MSG_WARNINGS_UPLOADFILETOOBIG="The selected file cannot be uploaded as it is bigger than the maximum upload size."
+COM_INSTALLER_MSG_WARNINGS_NOTSUPPORTEDDATABASEVERSIONDESC="The version of the database system you are using does not meet the <a href=\"https://downloads.joomla.org/technical-requirements\" class=\"alert-link\" target=\"_blank\" rel=\"noopener noreferrer\">core minimum requirements</a>. Your current configuration is:<br><ul><li>Database type: %1$s </li><li>Database version: %2$s</li><li>Minimum database version: %3$s</li></ul>This means you are running Joomla on an untested and unsupported environment which may result in issues with the core and extensions. Joomla will not provide any updates until your database is upgraded to reach at least the minimum supported version."
```
Comment (Quy) [inline administrator/language/en-GB/com_installer.ini:207]:
```suggestion
COM_INSTALLER_MSG_WARNINGS_NOTSUPPORTEDDATABASEVERSIONDESC="The version of the database system you are using does not meet the <a href=\"https://downloads.joomla.org/technical-requirements\" class=\"alert-link\" target=\"_blank\" rel=\"noopener noreferrer\">core minimum requirements</a>. Your current configuration is:<ul><li>Database type: %1$s </li><li>Database version: %2$s</li><li>Minimum database version: %3$s</li></ul>This means you are running Joomla on an untested and unsupported environment which may result in issues with the core and extensions. Joomla will not provide any updates until your database is upgraded to reach at least the minimum supported version."
```

Comment (Quy) [inline administrator/language/en-GB/com_installer.ini]:
```suggestion
COM_INSTALLER_MSG_WARNINGS_NOTSUPPORTEDDATABASEVERSIONDESC="The version of the database system you are using does not meet the <a href=\"https://downloads.joomla.org/technical-requirements\" class=\"alert-link\" target=\"_blank\" rel=\"noopener noreferrer\">core minimum requirements</a>. Your current configuration is:<ul><li>Database type: %1$s</li><li>Database version: %2$s</li><li>Minimum database version: %3$s</li></ul>This means you are running Joomla on an untested and unsupported environment which may result in issues with the core and extensions. Joomla will not provide any updates until your database is upgraded to reach at least the minimum supported version."
```

**Comment by webgras** [comment]
I have tested this item :white_check_mark: successfully on 71290ba3c9dadf8a76b4c2c0c075354350fc935f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38799">issues.joomla.org/tracker/joomla-cms/38799</a>.</sub>

**Comment by laoneo** [comment]
This should go at the end into 4.3 and not 4.2 as it is a new feature.

**Comment by zero-24** [comment]
> This should go at the end into 4.3 and not 4.2 as it is a new feature.

It is by far nothing like a new feature and its awaiting progress since ages, its fixing a process flaw. My understading from @fancyFranci was that she is just awaiting tests here for this to be merged which we just got so this should be good to go into the next release which can also be 4.3 not sure from the top of my head whether we get another 4.2 release anyway :)

**Comment by fancyFranci** [comment]
Yes, I'm finding testers for this because I want it in 4.2 :) The current message is misleading and kind of incorrect. The PR is fixing that problem.

**Comment by laoneo** [comment]
Why do you want it in 4.2? What is so urgent? This is far more than a bug fix and should not come into a patch release.

**Comment by zero-24** [comment]
Will close it here for now, looks like a none issue to people other than me.

**Comment by brianteeman** [comment]
This should not have been closed

**Pull request closed without merging**